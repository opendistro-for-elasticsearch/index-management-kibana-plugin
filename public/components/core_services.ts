import { createContext } from "react";
import { CoreStart } from "kibana/public";

const CoreServicesContext = createContext<CoreStart | null>(null);

const CoreServicesConsumer = CoreServicesContext.Consumer;

export { CoreServicesContext, CoreServicesConsumer };
