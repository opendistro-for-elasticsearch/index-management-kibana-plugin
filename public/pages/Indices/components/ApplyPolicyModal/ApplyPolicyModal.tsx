/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { Component, Fragment } from "react";
import { toastNotifications } from "ui/notify";
import _ from "lodash";
import {
  EuiButton,
  EuiButtonEmpty,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiOverlayMask,
  EuiComboBox,
  EuiFormRow,
  EuiFieldText,
  EuiCallOut,
  EuiText,
  EuiSpacer,
  EuiCodeBlock,
  EuiLink,
  EuiIcon,
} from "@elastic/eui";
import { BrowserServices } from "../../../../models/interfaces";
import { PolicyOption } from "../../models/interfaces";
import { Policy, State } from "../../../../../models/interfaces";
import { getErrorMessage } from "../../../../utils/helpers";
import { DOCUMENTATION_URL } from "../../../../utils/constants";

interface ApplyPolicyModalProps {
  onClose: () => void;
  services: BrowserServices;
  indices: string[];
}

interface ApplyPolicyModalState {
  isLoading: boolean;
  selectedPolicy: PolicyOption | null;
  selectedPolicyError: string;
  hasRolloverAction: boolean;
  policyOptions: PolicyOption[];
  rolloverAlias: string;
  rolloverAliasError: string;
  hasSubmitted: boolean;
}

export default class ApplyPolicyModal extends Component<ApplyPolicyModalProps, ApplyPolicyModalState> {
  state: ApplyPolicyModalState = {
    isLoading: false,
    selectedPolicy: null,
    selectedPolicyError: "",
    hasRolloverAction: false,
    policyOptions: [],
    rolloverAlias: "",
    rolloverAliasError: "",
    hasSubmitted: false,
  };

  async componentDidMount(): Promise<void> {
    await this.onPolicySearchChange("");
  }

  onApplyPolicy = async (selectedPolicy: PolicyOption, hasRolloverAction: boolean, rolloverAlias: string): Promise<void> => {
    try {
      const {
        onClose,
        indices,
        services: { indexService },
      } = this.props;
      const policyId = selectedPolicy.label;
      const applyPolicyResponse = await indexService.applyPolicy(indices, policyId);
      if (applyPolicyResponse.ok) {
        const { updatedIndices, failedIndices, failures } = applyPolicyResponse.response;
        if (updatedIndices) {
          toastNotifications.addSuccess(`Applied policy to ${updatedIndices} indices`);
          if (hasRolloverAction && rolloverAlias && indices.length === 1) {
            await this.onEditRolloverAlias(indices[0], rolloverAlias);
          }
        }
        if (failures) {
          toastNotifications.addDanger(
            `Failed to apply policy to ${failedIndices.map(failedIndex => `[${failedIndex.indexName}, ${failedIndex.reason}]`).join(", ")}`
          );
        }
        onClose();
      } else {
        toastNotifications.addDanger(applyPolicyResponse.error);
      }
    } catch (err) {
      toastNotifications.addDanger(getErrorMessage(err, "There was a problem adding policy to indices"));
    }
  };

  onEditRolloverAlias = async (index: string, rolloverAlias: string): Promise<void> => {
    const {
      services: { indexService },
    } = this.props;
    try {
      const response = await indexService.editRolloverAlias(index, rolloverAlias);
      if (response.ok) {
        if (response.response.acknowledged) {
          toastNotifications.addSuccess(`Edited rollover alias on ${index}`);
        } else {
          toastNotifications.addDanger(`Failed to edit rollover alias on ${index}`);
        }
      } else {
        toastNotifications.addDanger(response.error);
      }
    } catch (err) {
      toastNotifications.addDanger(getErrorMessage(err, `There was a problem editing rollover alias on ${index}`));
    }
  };

  onPolicySearchChange = async (searchValue: string): Promise<void> => {
    const {
      services: { indexService },
    } = this.props;
    this.setState({ isLoading: true, policyOptions: [] });
    try {
      const searchPoliciesResponse = await indexService.searchPolicies(searchValue, true);
      if (searchPoliciesResponse.ok) {
        const policies = searchPoliciesResponse.response.hits.hits.map((hit: { _id: string; _source: { policy: Policy } }) => ({
          label: hit._id,
          policy: hit._source.policy,
        }));
        this.setState({ policyOptions: policies });
      } else {
        if (searchPoliciesResponse.error.startsWith("[index_not_found_exception]")) {
          toastNotifications.addDanger("You have not created a policy yet");
        } else {
          toastNotifications.addDanger(searchPoliciesResponse.error);
        }
      }
    } catch (err) {
      toastNotifications.addDanger(err.message);
    }

    this.setState({ isLoading: false });
  };

  onChangeSelectedPolicy = (selectedPolicies: PolicyOption[]): void => {
    const selectedPolicy = selectedPolicies.length ? selectedPolicies[0] : null;
    const hasRolloverAction = this.hasRolloverAction(selectedPolicy);
    this.setState({
      selectedPolicy,
      hasRolloverAction,
      selectedPolicyError: this.getSelectedPolicyError(selectedPolicy),
    });
  };

  onChangeRolloverAlias = (e: React.ChangeEvent<HTMLInputElement>): void => {
    const rolloverAlias = e.target.value;
    this.setState({
      rolloverAlias,
      rolloverAliasError: this.getRolloverAliasError(rolloverAlias),
    });
  };

  onSubmit = async (): Promise<void> => {
    const { selectedPolicy, rolloverAlias } = this.state;
    const selectedPolicyError = this.getSelectedPolicyError(selectedPolicy);
    const rolloverAliasError = this.getRolloverAliasError(rolloverAlias);
    const hasSubmitError = !!selectedPolicyError || !!rolloverAliasError;
    if (hasSubmitError) {
      this.setState({ selectedPolicyError, rolloverAliasError, hasSubmitted: true });
    } else {
      // @ts-ignore
      await this.onApplyPolicy(selectedPolicy, this.hasRolloverAction(selectedPolicy), rolloverAlias);
    }
  };

  getRolloverAliasError = (rolloverAlias: string): string => {
    const { hasRolloverAction } = this.state;
    const { indices } = this.props;
    const hasSingleIndexSelected = indices.length === 1;
    const requiresAlias = hasRolloverAction && hasSingleIndexSelected;
    const hasAliasError = requiresAlias && !rolloverAlias;
    return hasAliasError ? "Required" : "";
  };

  getSelectedPolicyError = (selectedPolicy: PolicyOption | null): string => (selectedPolicy ? "" : "You must select a policy");

  hasRolloverAction = (selectedPolicy: PolicyOption | null): boolean =>
    _.get(selectedPolicy, "policy.states", []).some((state: State) => state.actions.some(action => action.hasOwnProperty("rollover")));

  renderRollover = (): React.ReactNode | null => {
    const { rolloverAlias, hasRolloverAction, rolloverAliasError, hasSubmitted } = this.state;
    const { indices } = this.props;
    const hasSingleIndexSelected = indices.length === 1;

    if (!hasRolloverAction) return null;

    if (hasSingleIndexSelected) {
      return (
        <EuiFormRow
          label="Rollover alias"
          helpText={
            <EuiText size="xs" grow={false}>
              <p>
                This policy includes a rollover action.
                Specify a rollover alias.{" "}
                <EuiLink href={DOCUMENTATION_URL} target="_blank">
                  Learn more <EuiIcon type="popout" size="s" />
                </EuiLink>
              </p>
            </EuiText>
          }
          isInvalid={hasSubmitted && !!rolloverAliasError}
          error={rolloverAliasError}
          fullWidth
        >
          <EuiFieldText
            isInvalid={hasSubmitted && !!rolloverAliasError}
            placeholder="Rollover alias"
            value={rolloverAlias}
            onChange={this.onChangeRolloverAlias}
            fullWidth
          />
        </EuiFormRow>
      );
    }

    return (
      <EuiCallOut
        style={{ width: "350px" }}
        title={
          <p>
            You are applying a policy with rollover to multiple indices. You will need to add a unique rollover_alias setting to each index.
          </p>
        }
        iconType="alert"
        size="s"
        color="warning"
      />
    );
  };

  renderPreview = (): React.ReactNode | null => {
    const { selectedPolicy } = this.state;
    if (!selectedPolicy) return null;

    let policyString = '';
    try {
      policyString = JSON.stringify({ policy: selectedPolicy.policy }, null, 4);
    } catch (err) {
      console.error(err);
    }

    if (!policyString) {
      return null;
    }

    return (
      <Fragment>
        <EuiText size="xs" grow={false}>
          <p><strong>Preview</strong></p>
        </EuiText>
        <EuiSpacer size="s" />
        <EuiCodeBlock language="json" fontSize="m" style={{ height: "200px" }}>
          {policyString}
        </EuiCodeBlock>
        <EuiSpacer size="m" />
      </Fragment>
    );
  };

  render() {
    const { policyOptions, selectedPolicy, selectedPolicyError, isLoading, hasSubmitted } = this.state;
    const { onClose } = this.props;
    const selectedOptions = selectedPolicy ? [selectedPolicy] : [];
    return (
      <EuiOverlayMask>
        {/*
            // @ts-ignore */}
        <EuiModal onCancel={onClose} onClose={onClose}>
          <EuiModalHeader>
            <EuiModalHeaderTitle>Apply policy</EuiModalHeaderTitle>
          </EuiModalHeader>

          <EuiModalBody>
            <EuiText size="xs" grow={false}>
              <p>
                Choose the policy you want to use for the selected indices.
                A copy of the policy will be created and applied to the indices.
              </p>
            </EuiText>
            <EuiSpacer size="m" />
            <EuiFormRow
              label="Policy ID"
              isInvalid={hasSubmitted && !!selectedPolicyError}
              error={selectedPolicyError}
              fullWidth
            >
              <EuiComboBox
                placeholder="Search policies"
                async
                options={policyOptions}
                singleSelection={{ asPlainText: true }}
                selectedOptions={selectedOptions}
                isLoading={isLoading}
                isInvalid={hasSubmitted && !!selectedPolicyError}
                onChange={this.onChangeSelectedPolicy}
                onSearchChange={this.onPolicySearchChange}
                fullWidth
              />
            </EuiFormRow>
            {this.renderPreview()}
            {this.renderRollover()}
          </EuiModalBody>

          <EuiModalFooter>
            <EuiButtonEmpty onClick={onClose} data-test-subj="applyPolicyModalCloseButton">
              Cancel
            </EuiButtonEmpty>

            <EuiButton onClick={this.onSubmit} fill data-test-subj="applyPolicyModalEditButton">
              Apply
            </EuiButton>
          </EuiModalFooter>
        </EuiModal>
      </EuiOverlayMask>
    );
  }
}
