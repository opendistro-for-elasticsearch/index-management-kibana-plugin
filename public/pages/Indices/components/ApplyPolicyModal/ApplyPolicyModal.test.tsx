/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, wait } from "@testing-library/react";
import ApplyPolicyModal from "./ApplyPolicyModal";
import { browserServicesMock, coreServicesMock, httpClientMock } from "../../../../../test/mocks";
import { CoreServicesContext } from "../../../../components/core_services";

// TODO: fireEvent for applyPolicy, but need to figure out how to get react-testing-library to work
//  with the combo_box in modal

describe("<ApplyPolicyModal /> spec", () => {
  it("renders the component", async () => {
    httpClientMock.post = jest.fn().mockResolvedValue({ ok: true, resp: { hits: { hits: [{ _id: "test_index" }] } } });
    render(<ApplyPolicyModal onClose={() => {}} services={browserServicesMock} indices={[]} />);

    // EuiOverlayMask appends an element to the body so we should have two, an empty div from react-test-library
    // and our EuiOverlayMask element
    expect(document.body.children).toHaveLength(2);
    expect(document.body.children[1]).toMatchSnapshot();
  });

  it("successfully calls search policies on mount", async () => {
    httpClientMock.post = jest.fn().mockResolvedValue({ ok: true, resp: { hits: { hits: [{ _id: "test_index" }] } } });
    const spy = jest.spyOn(browserServicesMock.indexService, "searchPolicies");
    render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ApplyPolicyModal onClose={() => {}} services={browserServicesMock} indices={[]} />
      </CoreServicesContext.Provider>
    );

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith("", true);
    expect(coreServicesMock.notifications.toasts.addDanger).not.toHaveBeenCalled();
  });

  it("adds danger toaster on safe error", async () => {
    httpClientMock.post = jest.fn().mockResolvedValue({ ok: false, error: "some error" });
    const spy = jest.spyOn(browserServicesMock.indexService, "searchPolicies");
    render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ApplyPolicyModal onClose={() => {}} services={browserServicesMock} indices={[]} />
      </CoreServicesContext.Provider>
    );

    // wait 1 tick for the searchPolicies promise to resolve
    await wait();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith("", true);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("some error");
  });

  it("adds danger toaster on unsafe error", async () => {
    httpClientMock.post = jest.fn().mockRejectedValue(new Error("testing error"));
    const spy = jest.spyOn(browserServicesMock.indexService, "searchPolicies");
    render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ApplyPolicyModal onClose={() => {}} services={browserServicesMock} indices={[]} />
      </CoreServicesContext.Provider>
    );

    // wait 1 tick for the searchPolicies promise to resolve
    await wait();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith("", true);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("testing error");
  });
});
