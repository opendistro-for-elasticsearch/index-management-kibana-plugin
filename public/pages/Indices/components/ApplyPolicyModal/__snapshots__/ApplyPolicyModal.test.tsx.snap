// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ApplyPolicyModal /> spec renders the component 1`] = `
<div
  class="euiOverlayMask euiOverlayMask--aboveHeader"
>
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="1"
  />
  <div
    data-focus-lock-disabled="false"
  >
    <div
      class="euiModal euiModal--maxWidth-default"
      tabindex="0"
    >
      <button
        aria-label="Closes this modal window"
        class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
        type="button"
      >
        EuiIconMock
      </button>
      <div
        class="euiModal__flex"
      >
        <div
          class="euiModalHeader"
        >
          <div
            class="euiModalHeader__title"
          >
            Apply policy
          </div>
        </div>
        <div
          class="euiModalBody"
        >
          <div
            class="euiModalBody__overflow"
          >
            <div
              class="euiText euiText--extraSmall euiText--constrainedWidth"
            >
              <p>
                Choose the policy you want to use for the selected indices. A copy of the policy will be created and applied to the indices.
              </p>
            </div>
            <div
              class="euiSpacer euiSpacer--m"
            />
            <div
              class="euiFormRow euiFormRow--fullWidth"
              id="some_html_id-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  aria-invalid="false"
                  class="euiFormLabel euiFormRow__label"
                  for="some_html_id"
                >
                  Policy ID
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  class="euiComboBox euiComboBox--fullWidth"
                  role="combobox"
                >
                  <div
                    class="euiFormControlLayout euiFormControlLayout--fullWidth"
                  >
                    <div
                      class="euiFormControlLayout__childrenWrapper"
                    >
                      <div
                        class="euiComboBox__inputWrap euiComboBox__inputWrap--fullWidth euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isLoading euiComboBox__inputWrap-isClearable"
                        data-test-subj="comboBoxInput"
                        tabindex="-1"
                      >
                        <p
                          class="euiComboBoxPlaceholder"
                        >
                          Search policies
                        </p>
                        <div
                          class="euiComboBox__input"
                          style="font-size: 14px; display: inline-block;"
                        >
                          <input
                            aria-controls=""
                            data-test-subj="comboBoxSearchInput"
                            id="some_html_id"
                            role="textbox"
                            style="box-sizing: content-box; width: 2px;"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                      <div
                        class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                      >
                        <span
                          class="euiLoadingSpinner euiLoadingSpinner--medium"
                        />
                        <button
                          aria-label="Open list of options"
                          class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                          data-test-subj="comboBoxToggleListButton"
                          type="button"
                        >
                          EuiIconMock
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiModalFooter"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--primary"
            data-test-subj="applyPolicyModalCloseButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonEmpty__content"
            >
              <span
                class="euiButtonEmpty__text"
              >
                Cancel
              </span>
            </span>
          </button>
          <button
            class="euiButton euiButton--primary euiButton--fill"
            data-test-subj="applyPolicyModalEditButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButton__content"
            >
              <span
                class="euiButton__text"
              >
                Apply
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    data-aria-hidden="true"
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
</div>
`;
